File: android-compose/app/src/main/java/com/motchill/androidcompose/feature/search/SearchViewModel.kt
package com.motchill.androidcompose.feature.search

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.motchill.androidcompose.core.storage.LikedMovieStore
import com.motchill.androidcompose.data.repository.MotchillRepository
import com.motchill.androidcompose.domain.model.SearchChoice
import com.motchill.androidcompose.domain.model.SearchFacetOption
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

internal class SearchViewModel(
    private val repository: MotchillRepository,
    private val likedMovieStore: LikedMovieStore,
    private val initialQuery: String,
    private val initialSlug: String,
    private val initialLabel: String,
    private val initialLikedOnly: Boolean,
) : ViewModel() {
    private val _uiState = MutableStateFlow(
        SearchUiState(
            searchText = initialQuery,
            searchInputValue = initialQuery,
            showLikedOnly = initialLikedOnly,
        ),
    )
    val uiState: StateFlow<SearchUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    fun refresh() {
        loadPage(_uiState.value.pageNumber)
    }

    fun onSearchTextChanged(value: String) {
        _uiState.value = _uiState.value.withSearchInput(value)
    }

    fun submitSearch(value: String? = null) {
        val nextValue = (value ?: _uiState.value.searchInputValue).trim()
        _uiState.value = _uiState.value
            .withSearchInput(nextValue)
            .commitSearch()
        loadPage(1)
    }

    fun clearSearch() {
        _uiState.value = _uiState.value.clearSearchInput()
        loadPage(1)
    }

    fun selectCategory(option: SearchFacetOption?) {
        if (option == null || !option.hasId || option.id <= 0) {
            clearCategory()
            return
        }
        _uiState.value = _uiState.value.withCategory(option)
        loadPage(1)
    }

    fun selectCountry(option: SearchFacetOption?) {
        if (option == null || !option.hasId || option.id <= 0) {
            clearCountry()
            return
        }
        _uiState.value = _uiState.value.withCountry(option)
        loadPage(1)
    }

    fun selectTypeRaw(choice: SearchChoice?) {
        if (choice == null || choice.value.trim().isEmpty()) {
            clearTypeRaw()
            return
        }
        _uiState.value = _uiState.value.withTypeRaw(choice)
        loadPage(1)
    }

    fun selectYear(choice: SearchChoice?) {
        if (choice == null || choice.value.trim().isEmpty()) {
            clearYear()
            return
        }
        _uiState.value = _uiState.value.withYear(choice)
        loadPage(1)
    }

    fun selectOrderBy(value: String) {
        _uiState.value = _uiState.value.withOrderBy(value)
        loadPage(1)
    }

    fun toggleLikedOnly() {
        _uiState.value = _uiState.value.toggleLikedOnly()
        // Flutter keeps the local liked list as the source of truth and does not refetch here.
    }

    fun clearCategory() {
        _uiState.value = _uiState.value.clearCategory()
        loadPage(1)
    }

    fun clearCountry() {
        _uiState.value = _uiState.value.clearCountry()
        loadPage(1)
    }

    fun clearTypeRaw() {
        _uiState.value = _uiState.value.clearTypeRaw()
        loadPage(1)
    }

    fun clearYear() {
        _uiState.value = _uiState.value.clearYear()
        loadPage(1)
    }

    fun clearOrderBy() {
        _uiState.value = _uiState.value.clearOrderBy()
        loadPage(1)
    }

    fun goToPage(pageNumber: Int) {
        loadPage(pageNumber.coerceAtLeast(1))
    }

    fun clearFilters() {
        _uiState.value = _uiState.value.copy(
            selectedCategoryId = null,
            selectedCategoryLabel = "",
            selectedCountryId = null,
            selectedCountryLabel = "",
            selectedTypeRaw = "",
            selectedTypeLabel = "",
            selectedYear = "",
            selectedOrderBy = DEFAULT_ORDER_BY,
            pageNumber = 1,
        )
        loadPage(1)
    }

    private fun load() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.withLoading()
            runCatching {
                coroutineScope {
                    val filtersDeferred = async { repository.loadSearchFilters() }
                    val likedDeferred = async { likedMovieStore.loadMovies() }
                    val filters = filtersDeferred.await()
                    val likedMovies = likedDeferred.await()
                    val preset = filters.findPreset(initialSlug)
                    _uiState.value = _uiState.value
                        .withLoadedFilters(filters)
                        .withLikedMovies(likedMovies)
                        .applyPreset(
                            preset = preset,
                            fallbackLabel = initialLabel,
                            slug = initialSlug,
                        )
                    if (initialQuery.trim().isNotEmpty()) {
                        _uiState.value = _uiState.value
                            .withSearchInput(initialQuery)
                            .commitSearch()
                    }
                    loadPage(1)
                }
            }.onFailure { error ->
                _uiState.value = _uiState.value.withError(error.message ?: error::class.java.simpleName)
            }
        }
    }

    private fun loadPage(pageNumber: Int) {
        viewModelScope.launch {
            val state = _uiState.value.withLoading(isSearching = _uiState.value.records.isNotEmpty())
            _uiState.value = state

            runCatching {
                val results = repository.loadSearchResults(
                    categoryId = state.selectedCategoryId,
                    countryId = state.selectedCountryId,
                    typeRaw = state.selectedTypeRaw,
                    year = state.selectedYear,
                    orderBy = state.selectedOrderBy,
                    search = state.searchText,
                    pageNumber = pageNumber,
                )

                _uiState.value = state.withSearchResults(results, pageNumber)
            }.onFailure { error ->
                _uiState.value = _uiState.value.withError(error.message ?: error::class.java.simpleName)
            }
        }
    }

    companion object {
        fun factory(
            repository: MotchillRepository,
            likedMovieStore: LikedMovieStore,
            initialQuery: String,
            initialSlug: String,
            initialLabel: String,
            startLikedOnly: Boolean,
        ): ViewModelProvider.Factory {
            return object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T {
                    if (modelClass.isAssignableFrom(SearchViewModel::class.java)) {
                        return SearchViewModel(
                            repository = repository,
                            likedMovieStore = likedMovieStore,
                            initialQuery = initialQuery,
                            initialSlug = initialSlug,
                            initialLabel = initialLabel,
                            initialLikedOnly = startLikedOnly,
                        ) as T
                    }
                    throw IllegalArgumentException("Unknown ViewModel class: ${modelClass.name}")
                }
            }
        }
    }
}

File: swiftUI/Features/Search/SearchViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class SearchViewModel {
    @ObservationIgnored
    private let repository: PhucTvRepository
    @ObservationIgnored
    private let likedMovieStore: PhucTvLikedMovieStoring
    @ObservationIgnored
    private let routeInput: SearchRouteInput

    var uiState: SearchUIState

    init(
        repository: PhucTvRepository,
        likedMovieStore: PhucTvLikedMovieStoring,
        routeInput: SearchRouteInput = SearchRouteInput(),
        uiState: SearchUIState? = nil
    ) {
        self.repository = repository
        self.likedMovieStore = likedMovieStore
        self.routeInput = routeInput
        self.uiState = uiState ?? SearchUIState(
            searchText: routeInput.initialQuery,
            searchInputValue: routeInput.initialQuery,
            showLikedOnly: routeInput.startLikedOnly
        )
    }

    func load() async {
        uiState = uiState.withLoading()

        do {
            async let filtersTask = repository.loadSearchFilters()
            async let likedTask = likedMovieStore.loadMovies()
            let filters = try await filtersTask
            let likedMovies = try await likedTask
            let preset = filters.findPreset(slug: routeInput.presetSlug)

            uiState = uiState
                .withLoadedFilters(filters)
                .withLikedMovies(likedMovies)
                .applyPreset(
                    preset,
                    fallbackLabel: routeInput.initialLabel,
                    slug: routeInput.presetSlug
                )

            if !routeInput.initialQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                uiState = uiState
                    .withSearchInput(routeInput.initialQuery)
                    .commitSearch()
            }

            await loadPage(1)
        } catch {
            PhucTvLogger.shared.error(
                error,
                message: "Search load failed",
                metadata: ["state": "search_load"]
            )
            uiState = uiState.withError(error.localizedDescription)
        }
    }

    func refresh() async {
        if uiState.showLikedOnly {
            uiState = uiState.withIdle()
            return
        }
        await loadPage(uiState.pageNumber)
    }

    func onSearchTextChanged(_ value: String) {
        uiState = uiState.withSearchInput(value)
    }

    func submitSearch(_ value: String? = nil) async {
        let nextValue = (value ?? uiState.searchInputValue).trimmingCharacters(in: .whitespacesAndNewlines)
        uiState = uiState
            .withSearchInput(nextValue)
            .commitSearch()
        await loadPage(1)
    }

    func clearSearch() async {
        uiState = uiState.clearSearchInput()
        await loadPage(1)
    }

    func selectCategory(_ option: PhucTvSearchFacetOption?) async {
        if let option, option.hasID {
            uiState = uiState.withCategory(option)
        } else {
            uiState = uiState.clearCategory()
        }
        await loadPage(1)
    }

    func selectCountry(_ option: PhucTvSearchFacetOption?) async {
        if let option, option.hasID {
            uiState = uiState.withCountry(option)
        } else {
            uiState = uiState.clearCountry()
        }
        await loadPage(1)
    }

    func selectTypeRaw(_ option: PhucTvSearchChoice?) async {
        if let option, !option.value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            uiState = uiState.withTypeRaw(option)
        } else {
            uiState = uiState.clearTypeRaw()
        }
        await loadPage(1)
    }

    func selectYear(_ option: PhucTvSearchChoice?) async {
        if let option, !option.value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            uiState = uiState.withYear(option)
        } else {
            uiState = uiState.clearYear()
        }
        await loadPage(1)
    }

    func selectOrderBy(_ value: String) async {
        uiState = uiState.withOrderBy(value)
        await loadPage(1)
    }

    func toggleLikedOnly() {
        uiState = uiState.toggleLikedOnly()
    }

    func clearCategory() async {
        uiState = uiState.clearCategory()
        await loadPage(1)
    }

    func clearCountry() async {
        uiState = uiState.clearCountry()
        await loadPage(1)
    }

    func clearTypeRaw() async {
        uiState = uiState.clearTypeRaw()
        await loadPage(1)
    }

    func clearYear() async {
        uiState = uiState.clearYear()
        await loadPage(1)
    }

    func clearOrderBy() async {
        uiState = uiState.clearOrderBy()
        await loadPage(1)
    }

    func goToPage(_ pageNumber: Int) async {
        let target = max(pageNumber, 1)
        if uiState.showLikedOnly {
            uiState = uiState.withPageNumber(target)
            return
        }
        await loadPage(target)
    }

    func pickerOptions(for kind: SearchPickerKind) -> [SearchUIPickerOption] {
        switch kind {
        case .category:
            return uiState.filters.categoryOptionsWithAll().map {
                SearchUIPickerOption(
                    id: "category-\($0.id)-\($0.slug)",
                    title: $0.name,
                    subtitle: $0.slug,
                    isSelected: uiState.selectedCategoryID == $0.id || (!$0.hasID && uiState.selectedCategoryID == nil)
                )
            }
        case .country:
            return uiState.filters.countryOptionsWithAll().map {
                SearchUIPickerOption(
                    id: "country-\($0.id)-\($0.slug)",
                    title: $0.name,
                    subtitle: $0.slug,
                    isSelected: uiState.selectedCountryID == $0.id || (!$0.hasID && uiState.selectedCountryID == nil)
                )
            }
        case .type:
            return searchTypeOptions.map {
                SearchUIPickerOption(
                    id: "type-\($0.value)",
                    title: $0.label,
                    subtitle: $0.value,
                    isSelected: uiState.selectedTypeRaw == $0.value
                )
            }
        case .year:
            return searchYearOptions.map {
                SearchUIPickerOption(
                    id: "year-\($0.value)",
                    title: $0.label,
                    subtitle: $0.value,
                    isSelected: uiState.selectedYear == $0.value
                )
            }
        case .order:
            return searchOrderOptions.map {
                SearchUIPickerOption(
                    id: "order-\($0.value)",
                    title: $0.label,
                    subtitle: $0.value,
                    isSelected: uiState.selectedOrderBy == $0.value
                )
            }
        }
    }

    private func loadPage(_ pageNumber: Int) async {
        if uiState.showLikedOnly {
            uiState = uiState.withPageNumber(pageNumber).withIdle()
            return
        }

        let state = uiState.withLoading(isSearching: !uiState.records.isEmpty)
        uiState = state

        do {
            let results = try await repository.loadSearchResults(
                categoryId: state.selectedCategoryID,
                countryId: state.selectedCountryID,
                typeRaw: state.selectedTypeRaw,
                year: state.selectedYear,
                orderBy: state.selectedOrderBy,
                isChieuRap: false,
                is4k: false,
                search: state.searchText,
                pageNumber: pageNumber
            )
            uiState = state.withSearchResults(results, pageNumber: pageNumber)
        } catch {
            PhucTvLogger.shared.error(
                error,
                message: "Search page load failed",
                metadata: [
                    "page": "\(pageNumber)",
                    "query": state.searchText,
                ]
            )
            uiState = uiState.withError(error.localizedDescription)
        }
    }
}

